components/config_info/src/lib.rs

```diff
@@ -0,0 +1,330 @@
+/// misleading. In this can, user can explicitly set the `default_desc` attribue to privide a string to
+/// describe the default value choosen rule.
+///
+/// This same rule also applies for the `min_desc` and `max_desc`.
+///
+///
+/// # Config Value Options(#[config_info(opitons= [ .., .. ])])
+/// The `options` attribue define the set of all valid valid for target field. The value of each element
+/// must be either the same type with target field or can be converted to target type with `TryInto::try_into`.
+///
+///
+/// Example:
+/// Please refer the unit test at the bottom of this file.
+///
+pub trait ConfigInfo {
+    type Encoder: Serialize;
+    /// Get encoder that can be serialize with `serde::Serializer`
+    /// with the disappear of `#[config_info(skip)]` field
+    // NOTE: the `input` parameter is use to extract the value_in_file value,
+    // if the `--config` is not provide, just use self instead.
+    fn get_cfg_encoder(&self, input: &Self) -> Self::Encoder;
+}
+
+// this is the stub Serialize stuct for each config field
+#[doc(hidden)]
+#[derive(Serialize)]
+#[serde(default)]
+#[serde(rename_all = "PascalCase")]
+pub struct FieldInfo<T: Serialize> {
+    #[serde(rename = "Type")]
+    field_type: FieldCfgType,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    min_value: Option<ConfigValue<T>>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    max_value: Option<ConfigValue<T>>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    value_options: Option<Vec<T>>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    default_value: Option<ConfigValue<T>>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    value_in_file: Option<T>,
+    description: String,
+}
+
+impl<T: Serialize> FieldInfo<T> {
+    pub fn new(
+        field_type: FieldCfgType,
+        default_value: Option<ConfigValue<T>>,
+        value_in_file: Option<T>,
+        description: String,
+    ) -> Self {
+        Self {
+            field_type,
+            min_value: None,
+            max_value: None,
+            value_options: None,
+            default_value,
+            value_in_file,
+            description,
+        }
+    }
+
+    pub fn set_min_value(mut self, v: ConfigValue<T>) -> Self {
+        self.min_value = Some(v);
+        self
+    }
+
+    pub fn set_max_value(mut self, v: ConfigValue<T>) -> Self {
+        self.max_value = Some(v);
+        self
+    }
+
+    pub fn set_value_options(mut self, ops: Vec<T>) -> Self {
+        self.value_options = Some(ops);
+        self
+    }
+}
+
+#[derive(Serialize)]
+#[serde(untagged)]
+pub enum ConfigValue<T: Serialize> {
+    // a concrete value of type T
+    Concrete(T),
+    // a string description.
+    Desc(String),
+}
+
+#[derive(Serialize)]
+pub enum FieldCfgType {
+    /// represent `[T; N]` and `Vec<T>` types.
+    Array,
+    /// represent `bool` type
+    Boolean,
+    /// represent map types `HashMap<K, V>` or `BTreeMap<K, V>`
+    Map,
```

Comment:
What config uses this type?

Reply:
The only config is `server.labels` by now.